drizzle-orm/src/mysql-core/query-builders/set-operators.ts

```diff
@@ -0,0 +1,351 @@
+	type ValidateShape,
+	type ValueOrArray,
+} from '~/index.ts';
+import type {
+	MySqlSession,
+	PreparedQueryConfig,
+	PreparedQueryHKTBase,
+	PreparedQueryKind,
+} from '~/mysql-core/session.ts';
+import { TypedQueryBuilder } from '~/query-builders/query-builder.ts';
+import type {
+	BuildSubquerySelection,
+	JoinNullability,
+	SelectMode,
+	SelectResult,
+} from '~/query-builders/select.types.ts';
+import { type ColumnsSelection } from '~/view.ts';
+import type { MySqlColumn } from '../columns/common.ts';
+import type { MySqlDialect } from '../dialect.ts';
+import type { SubqueryWithSelection } from '../subquery.ts';
+
+type SetOperator = 'union' | 'intersect' | 'except';
+
+const getMySqlSetOperators = () => {
+	return {
+		union,
+		unionAll,
+		intersect,
+		intersectAll,
+		except,
+		exceptAll,
+	};
+};
+
+type MySqlSetOperators = ReturnType<typeof getMySqlSetOperators>;
+
+type SetOperatorRightSelect<
+	TValue extends TypedQueryBuilder<any, SelectResult<TSelection, TSelectMode, TNullabilityMap>[]>,
+	TSelection extends ColumnsSelection,
+	TSelectMode extends SelectMode,
+	TNullabilityMap extends Record<string, JoinNullability>,
+> = TValue extends MySqlSetOperatorBuilder<any, infer TSel, infer TMode, any, infer TNull> ? ValidateShape<
+		SelectResult<TSel, TMode, TNull>,
+		SelectResult<TSelection, TSelectMode, TNullabilityMap>,
+		TypedQueryBuilder<any, SelectResult<TSelection, TSelectMode, TNullabilityMap>[]>
+	>
+	: TValue;
+
+type SetOperatorRestSelect<
+	TValue extends readonly TypedQueryBuilder<any, any[]>[],
+	Valid,
+> = TValue extends [infer First, ...infer Rest]
+	? First extends MySqlSetOperatorBuilder<any, infer TSel, infer TMode, any, infer TNull>
+		? Rest extends TypedQueryBuilder<any, any[]>[] ? [
+				ValidateShape<SelectResult<TSel, TMode, TNull>, Valid, TValue[0]>,
+				...SetOperatorRestSelect<Rest, Valid>,
+			]
+		: ValidateShape<SelectResult<TSel, TMode, TNull>, Valid, TValue>
+	: never[]
+	: TValue;
+
+export interface MySqlSetOperatorConfig {
+	fields: Record<string, unknown>;
+	operator: SetOperator;
+	isAll: boolean;
+	leftSelect: MySqlSetOperatorBuilder<any, any, any, any, any>;
+	rightSelect: TypedQueryBuilder<any, any[]>;
+	limit?: number | Placeholder;
+	orderBy?: (MySqlColumn | SQL | SQL.Aliased)[];
+	offset?: number | Placeholder;
+}
+
+export interface MySqlSetOperatorBuilder<
+	TTableName extends string | undefined,
+	TSelection extends ColumnsSelection,
+	TSelectMode extends SelectMode,
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	TPreparedQueryHKT extends PreparedQueryHKTBase,
+	TNullabilityMap extends Record<string, JoinNullability> = TTableName extends string ? Record<TTableName, 'not-null'>
+		: {},
+> extends
+	TypedQueryBuilder<
+		BuildSubquerySelection<TSelection, TNullabilityMap>,
+		SelectResult<TSelection, TSelectMode, TNullabilityMap>[]
+	>,
+	QueryPromise<SelectResult<TSelection, TSelectMode, TNullabilityMap>[]>
+{}
+
+export abstract class MySqlSetOperatorBuilder<
```

Comment:
Why is this an abstract class? Are there multiple subclasses extending it?

Reply:
This abstract class is extended by `MySqlSelectQueryBuilderBase` and `MySqlSetOperator` concrete class. I decided this was a better approach than only a concrete class because there were subtle difference in some of the shared methods. Also, I'm avoiding a lot of changes for the `with recursive` implementation. This approach is working great, what do you think would be the benefit of a straight concrete class?

Reply:
Also, I just realized I made a mistake, and it's probably the reason you're wondering about why is this an abstract class. the `applyMixins` function is supposed to be applied to `MySqlSetOperator` and not the abstract `MySqlSetOperatorBuilder`